pkg/storegateway/indexheader/reader_pool_test.go

```diff
@@ -28,54 +28,108 @@ import (
 
 func TestReaderPool_NewBinaryReader(t *testing.T) {
 	tests := map[string]struct {
-		lazyReaderEnabled     bool
-		lazyReaderIdleTimeout time.Duration
+		lazyReaderEnabled                           bool
+		lazyReaderIdleTimeout                       time.Duration
+		eagerLoadReaderEnabled                      bool
+		initialSync                                 bool
+		createLazyLoadedHeadersSnapshotFn           func(blockId ulid.ULID) lazyLoadedHeadersSnapshot
+		expectedLoadCountMetricBeforeLabelNamesCall int
+		expectedLoadCountMetricAfterLabelNamesCall  int
 	}{
 		"lazy reader is disabled": {
-			lazyReaderEnabled: false,
+			lazyReaderEnabled:                          false,
+			expectedLoadCountMetricAfterLabelNamesCall: 0, // no lazy loading
 		},
 		"lazy reader is enabled but close on idle timeout is disabled": {
-			lazyReaderEnabled:     true,
-			lazyReaderIdleTimeout: 0,
+			lazyReaderEnabled:                          true,
+			lazyReaderIdleTimeout:                      0,
+			expectedLoadCountMetricAfterLabelNamesCall: 1,
 		},
 		"lazy reader and close on idle timeout are both enabled": {
-			lazyReaderEnabled:     true,
-			lazyReaderIdleTimeout: time.Minute,
+			lazyReaderEnabled:                          true,
+			lazyReaderIdleTimeout:                      time.Minute,
+			expectedLoadCountMetricAfterLabelNamesCall: 1,
+		},
+		"block is present in pre-shutdown loaded blocks and eager-loading is enabled during initial sync": {
```

Comment:
[nit] Perhaps this would be clearer? (it's ambiguous if the "during initial sync" is referring to when the eager-loading is enabled or when the index header is being used)
```suggestion
		"block is present in pre-shutdown loaded blocks and eager-loading is enabled, loading index header during initial sync": {
```